File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.swift
class Solution {
    
    func isRepeated(_ mp: [Character: Int]) -> Bool{
        for (key, val) in mp{
            if val > 1 {
                return true
            }
        }
        return false
    }
    
    func lengthOfLongestSubstring(_ s: String) -> Int {
        var inputArray = Array(s), windowStart = 0, mx = 0
        
        var mp = [Character: Int]()
        
        for windowEnd in 0..<inputArray.count{
            mp[inputArray[windowEnd], default: 0 ] += 1
            
            if  mp[inputArray[windowEnd], default: 0 ] <= 1{
               mx = max(mx, windowEnd - windowStart + 1) 
            }else{
                while(  mp[inputArray[windowEnd], default: 0 ] > 1){
                    mp[inputArray[windowStart], default: 0] -= 1
                    windowStart += 1
                }
            }
        }
        
        return mx
        
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> charMap;
        
        int windowStart = 0, count = 0, maxSubLength = 0, size = s.size();
        
        for(int windowEnd = 0; windowEnd < size; windowEnd++){
             int charIndex = s[windowEnd];
             charMap[charIndex] ++ ;
             
             while(charMap[charIndex] > 1){
                 int newCharIndex = s[windowStart] ;
                 charMap[newCharIndex] -- ;
                 windowStart ++ ;
             }
            
            maxSubLength = max(maxSubLength, windowEnd - windowStart + 1);
        }
        
        return maxSubLength;
    }
};